### PR TITLE
Added Test_protect to prevent cascading failures. Fixed callback func…

### PR DESCRIPTION
**Pull request by lundinc2**
…tion signature. Fixed test case for master not initialized API return. Added config for frequency

Description
-----------

Checklist:
----------

- [ ] I have tested my changes. No regression in existing tests.
- [ ] My code is Linted.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by lundinc2** [comment]
TODO: Run uncrustify after this PR gets approved.

File: libraries/abstractions/common_io/test/test_iot_spi.c [hunk 1]
```diff
@@ -102,12 +106,12 @@ TEST_GROUP_RUNNER( TEST_IOT_SPI )
...
     RUN_TEST_CASE( TEST_IOT_SPI, AFQP_IotSPI_CancelSuccess );
 
-    //RUN_TEST_CASE( TEST_IOT_SPI, AFQP_IotSPI_ReadSync );
-    //RUN_TEST_CASE( TEST_IOT_SPI, AFQP_IotSPI_ReadAsync );
-    //RUN_TEST_CASE( TEST_IOT_SPI, AFQP_IotSPI_WriteSync );
-    //RUN_TEST_CASE( TEST_IOT_SPI, AFQP_IotSPI_WriteAsync );
-    //RUN_TEST_CASE( TEST_IOT_SPI, AFQP_IotSPI_TransferSync );
-    //RUN_TEST_CASE( TEST_IOT_SPI, AFQP_IotSPI_TransferAsync );
+    RUN_TEST_CASE( TEST_IOT_SPI, AFQP_IotSPI_ReadSync );
+    RUN_TEST_CASE( TEST_IOT_SPI, AFQP_IotSPI_ReadAsync );
```
Comment (vjjunnut) [inline libraries/abstractions/common_io/test/test_iot_spi.c]: These tests require a SPI slave, and are run with assisted test framework. By enabling these here, how do we expect them to run?

File: libraries/abstractions/common_io/test/test_iot_spi.c [hunk 2]
```diff
@@ -200,25 +208,29 @@ TEST( TEST_IOT_SPI, AFQP_IotSPI_ReadSync )
...
+        lRetVal = iot_spi_ioctl( xSPIHandle, eSPISetMasterConfig, &xTestConfig);
+        TEST_ASSERT_EQUAL( IOT_SPI_SUCCESS, lRetVal );
 
-    //lRetVal = iot_spi_ioctl( xSPIHandle, eSPIGetRxNoOfbytes, &xBytesRx);
-    //TEST_ASSERT_EQUAL( IOT_SPI_SUCCESS, lRetVal );
-    //TEST_ASSERT_EQUAL( xBytesRx, 4 );
+        lRetVal = iot_spi_read_sync( xSPIHandle, ucRxBuf, 4);
+        TEST_ASSERT_EQUAL( IOT_SPI_SUCCESS, lRetVal );
+
+        lRetVal = iot_spi_ioctl( xSPIHandle, eSPIGetRxNoOfbytes, &xBytesRx);
```
Comment (vjjunnut) [inline libraries/abstractions/common_io/test/test_iot_spi.c]: for sync APIs, RxNoOfBytes may not return the bytes. Sync API shall return success only if all the bytes are available.

File: libraries/abstractions/common_io/test/test_iot_spi.c [hunk 3]
```diff
@@ -31,65 +31,69 @@
...
 /*-----------------------------------------------------------*/
 
 /**
  * @brief Setup function called before each test in this group is executed.
  */
 TEST_SETUP( TEST_IOT_SPI )
 {
-    xtestIotSpiSemaphore = xSemaphoreCreateBinary();
-    TEST_ASSERT_NOT_EQUAL(NULL, xtestIotSpiSemaphore);
+   xtestIotSPISemaphore = xSemaphoreCreateCountingStatic( 10, 0, &xtestIotSPICompleted );
```
Comment (brianfr-amazon) [inline libraries/abstractions/common_io/test/test_iot_spi.c]: Why the change?

File: libraries/abstractions/common_io/test/test_iot_spi.c [hunk 4]
```diff
@@ -308,20 +327,24 @@ TEST( TEST_IOT_SPI, AFQP_IotSPI_WriteSync )
...
 
-    //lRetVal = iot_spi_ioctl( xSPIHandle, eSPIGetTxNoOfbytes, &xBytesTx);
-    //TEST_ASSERT_EQUAL( IOT_SPI_SUCCESS, lRetVal );
-    //TEST_ASSERT_EQUAL( xBytesTx, 4 );
+        lRetVal = iot_spi_write_sync( xSPIHandle, ucTxBuf, 4);
+        TEST_ASSERT_EQUAL( IOT_SPI_SUCCESS, lRetVal );
+
+        lRetVal = iot_spi_ioctl( xSPIHandle, eSPIGetTxNoOfbytes, &xBytesTx);
+        TEST_ASSERT_EQUAL( IOT_SPI_SUCCESS, lRetVal );
+        TEST_ASSERT_EQUAL( xBytesTx, 4 );
```
Comment (vjjunnut) [inline libraries/abstractions/common_io/test/test_iot_spi.c]: Same here. for sync APIs, TxNoOfBytes may not return number of bytes transmitted. Its used for async calls. Sync API is supposed to return SUCCESS only if all the bytes are sent.

File: libraries/abstractions/common_io/test/test_iot_spi.c [hunk 5]
```diff
@@ -681,7 +755,7 @@ TEST( TEST_IOT_SPI, AFQP_IotSPI_ReadSyncFuzzing )
     TEST_ASSERT_NOT_EQUAL( NULL, xSPIHandle );
 
     lRetVal = iot_spi_read_sync(NULL, ucBuffer, SPI_BUFFER_SIZE);
-    TEST_ASSERT_EQUAL( IOT_SPI_INVALID_VALUE, lRetVal );
```
Comment (brianfr-amazon) [inline libraries/abstractions/common_io/test/test_iot_spi.c]: Following the same behavior of all the other api's, if we pass in a NULL handle the return is INVALID_VALUE, not NOT_INITIALIZED.

File: libraries/abstractions/common_io/test/test_iot_spi.c (hunk 3 above)
Comment (vjjunnut) [inline libraries/abstractions/common_io/test/test_iot_spi.c]: Whats the advantage of using counting here?

File: libraries/abstractions/common_io/test/test_iot_spi.c (hunk 1 above)
Comment (lundinc2) [reply libraries/abstractions/common_io/test/test_iot_spi.c]: I will add the #define config so they can be toggled on or off.

File: libraries/abstractions/common_io/test/test_iot_spi.c (hunk 3 above)
Comment (lundinc2) [reply libraries/abstractions/common_io/test/test_iot_spi.c]: I was debugging a strange behavior, fixed that now. I'm okay with changing it back to binary, but I would like to keep it static if you agree

File: libraries/abstractions/common_io/test/test_iot_spi.c (hunk 4 above)
Comment (lundinc2) [reply libraries/abstractions/common_io/test/test_iot_spi.c]: Why can't it be used with sync? What if a partial amount of the operation was sent and we wanted to see how much of our data was able send?

File: libraries/abstractions/common_io/test/test_iot_spi.c (hunk 5 above)
Comment (lundinc2) [reply libraries/abstractions/common_io/test/test_iot_spi.c]: I agree with this, what is the purpose for IOT_SPI_MASTER_NOT_INITIALIZED in the API? Should we just remove it if we are not going to use it?

File: libraries/abstractions/common_io/test/test_iot_spi.c (hunk 3 above)
Comment (lundinc2) [reply libraries/abstractions/common_io/test/test_iot_spi.c]: I will revert this to binary
